File: static/js/matcher.js
/**
 * Word matching algorithm for OCR evaluation.
 * Implements two-phase matching: exact matching followed by fuzzy matching.
 */

/**
 * Count occurrences of items in an array (like Python's Counter).
 *
 * @param {Array} arr - Array of items
 * @returns {Object} Object with items as keys and counts as values
 */
function counter(arr) {
    const counts = {};
    for (const item of arr) {
        counts[item] = (counts[item] || 0) + 1;
    }
    return counts;
}

/**
 * Match OCR words to ground truth words using exact matching only.
 * Order-invariant, handles word counts properly.
 *
 * Algorithm:
 * 1. Exact Matching: Match identical words
 * 2. Mark Unmatched: Remaining words are gt_only or ocr_only
 *
 * @param {Array<string>} gtWords - List of preprocessed ground truth words
 * @param {Array<string>} ocrWords - List of preprocessed OCR output words
 * @returns {Array} List of match tuples [gtWord, ocrWord, editDistance, matchType]
 */
function matchWords(gtWords, ocrWords) {
    const matches = [];

    // Count word occurrences (handles duplicates)
    const gtCounts = counter(gtWords);
    const ocrCounts = counter(ocrWords);

    // Track matched word instances
    const gtMatched = {};
    const ocrMatched = {};

    // PHASE 1: Exact matching (greedy, highest counts first)
    const gtWordsUnique = Object.keys(gtCounts).sort((a, b) => gtCounts[b] - gtCounts[a]);

    for (const gtWord of gtWordsUnique) {
        if (gtWord in ocrCounts) {
            // Match as many instances as possible
            const matchCount = Math.min(gtCounts[gtWord], ocrCounts[gtWord]);
            for (let i = 0; i < matchCount; i++) {
                matches.push([gtWord, gtWord, 0, 'exact']);
            }
            gtMatched[gtWord] = matchCount;
            ocrMatched[gtWord] = matchCount;
        }
    }

    // PHASE 2: Mark unmatched words
    // Create lists of remaining unmatched instances
    const gtRemaining = [];
    for (const [word, count] of Object.entries(gtCounts)) {
        const matched = gtMatched[word] || 0;
        for (let i = 0; i < count - matched; i++) {
            gtRemaining.push(word);
        }
    }

    const ocrRemaining = [];
    for (const [word, count] of Object.entries(ocrCounts)) {
        const matched = ocrMatched[word] || 0;
        for (let i = 0; i < count - matched; i++) {
            ocrRemaining.push(word);
        }
    }

    // GT words with no match (false negatives)
    for (const gtWord of gtRemaining) {
        matches.push([gtWord, null, null, 'gt_only']);
    }

    // OCR words with no match (false positives)
    for (const ocrWord of ocrRemaining) {
        matches.push([null, ocrWord, null, 'ocr_only']);
    }

    return matches;
}

/**
 * Create annotations for text visualization, preserving original word order.
 *
 * @param {Array} wordData - List of word data objects with keys: normalized, original, position
 * @param {Array} matches - Match results from matchWords()
 * @param {boolean} isGroundTruth - True if annotating GT text, False if OCR text
 * @returns {Array} List of annotation objects
 */
function createAnnotations(wordData, matches, isGroundTruth = true) {
    // Build a mapping from normalized words to their match information
    // Handle duplicates by tracking available matches
    const matchMap = {}; // normalized_word -> list of {matchInfo, used: bool}

    if (isGroundTruth) {
        for (let idx = 0; idx < matches.length; idx++) {
            const [gtWord, ocrWord, editDist, matchType] = matches[idx];
            if (['exact', 'gt_only'].includes(matchType)) {
                if (!(gtWord in matchMap)) {
                    matchMap[gtWord] = [];
                }
                matchMap[gtWord].push({
                    matched_with: ocrWord,
                    edit_distance: editDist,
                    match_type: matchType,
                    match_id: matchType === 'exact' ? `match_${idx}` : null,
                    used: false  // Track if this match has been consumed
                });
            }
        }
    } else {
        for (let idx = 0; idx < matches.length; idx++) {
            const [gtWord, ocrWord, editDist, matchType] = matches[idx];
            if (['exact', 'ocr_only'].includes(matchType)) {
                if (!(ocrWord in matchMap)) {
                    matchMap[ocrWord] = [];
                }
                matchMap[ocrWord].push({
                    matched_with: gtWord,
                    edit_distance: editDist,
                    match_type: matchType,
                    match_id: matchType === 'exact' ? `match_${idx}` : null,
                    used: false  // Track if this match has been consumed
                });
            }
        }
    }

    // Create annotations in original word order
    const annotations = [];

    for (const wordInfo of wordData) {
        const normalized = wordInfo.normalized;
        const original = wordInfo.original;

        if (normalized in matchMap) {
            // Find the next available (unused) match for this word
            // Prefer exact matches, then fuzzy, then only matches
            const availableMatches = matchMap[normalized].filter(m => !m.used);

            let matchInfo;
            if (availableMatches.length > 0) {
                // Sort by match quality: exact > only
                const matchPriority = {'exact': 0, 'gt_only': 1, 'ocr_only': 1};
                availableMatches.sort((a, b) =>
                    (matchPriority[a.match_type] || 2) - (matchPriority[b.match_type] || 2)
                );

                // Use the best available match
                matchInfo = availableMatches[0];
                matchInfo.used = true; // Mark as consumed
            } else {
                // All matches for this word have been used
                // Use the last match (reuse it)
                matchInfo = matchMap[normalized][matchMap[normalized].length - 1];
            }

            annotations.push({
                word: original,  // Use original word, not normalized
                match_type: matchInfo.match_type,
                matched_with: matchInfo.matched_with,
                edit_distance: matchInfo.edit_distance,
                match_id: matchInfo.match_id
            });
        } else {
            // No match found - shouldn't happen but handle gracefully
            annotations.push({
                word: original,
                match_type: isGroundTruth ? 'gt_only' : 'ocr_only',
                matched_with: null,
                edit_distance: null,
                match_id: null
            });
        }
    }

    return annotations;
}

File: core/matcher.py
"""
Word matching algorithm for OCR evaluation.
Implements two-phase matching: exact matching followed by fuzzy matching.
"""

from collections import Counter
from .utils import levenshtein_distance


def match_words(gt_words, ocr_words, threshold=1):
    """
    Match OCR words to ground truth words using exact and fuzzy matching.
    Order-invariant, handles word counts properly.

    Algorithm:
    1. Phase 1 - Exact Matching: Match identical words first
    2. Phase 2 - Fuzzy Matching: Match remaining words using edit distance ≤ threshold

    Args:
        gt_words (list): List of preprocessed ground truth words
        ocr_words (list): List of preprocessed OCR output words
        threshold (int): Maximum edit distance for fuzzy matching (default: 1)

    Returns:
        list: List of match tuples (gt_word, ocr_word, edit_distance, match_type)
              match_type can be: 'exact', 'fuzzy', 'gt_only', 'ocr_only'
    """
    matches = []

    # Count word occurrences (handles duplicates)
    gt_counts = Counter(gt_words)
    ocr_counts = Counter(ocr_words)

    # Track matched word instances
    gt_matched = Counter()
    ocr_matched = Counter()

    # PHASE 1: Exact matching (greedy, highest counts first)
    # This ensures exact matches are prioritized
    for gt_word in sorted(gt_counts.keys(), key=lambda w: gt_counts[w], reverse=True):
        if gt_word in ocr_counts:
            # Match as many instances as possible
            match_count = min(gt_counts[gt_word], ocr_counts[gt_word])
            for _ in range(match_count):
                matches.append((gt_word, gt_word, 0, 'exact'))
            gt_matched[gt_word] = match_count
            ocr_matched[gt_word] = match_count

    # PHASE 2: Fuzzy matching for unmatched words
    # Create lists of remaining unmatched instances
    gt_remaining = []
    for word, count in gt_counts.items():
        matched = gt_matched.get(word, 0)
        gt_remaining.extend([word] * (count - matched))

    ocr_remaining = []
    for word, count in ocr_counts.items():
        matched = ocr_matched.get(word, 0)
        ocr_remaining.extend([word] * (count - matched))

    # Greedy fuzzy matching: find best pairs below threshold
    while gt_remaining and ocr_remaining:
        best_match = None
        best_distance = threshold + 1
        best_gt_idx = None
        best_ocr_idx = None

        # Find the best match (lowest edit distance ≤ threshold)
        for gt_idx, gt_word in enumerate(gt_remaining):
            for ocr_idx, ocr_word in enumerate(ocr_remaining):
                distance = levenshtein_distance(gt_word, ocr_word)

                # Check if this is a better match
                if distance <= threshold and distance < best_distance:
                    best_distance = distance
                    best_match = (gt_word, ocr_word, distance)
                    best_gt_idx = gt_idx
                    best_ocr_idx = ocr_idx

        # If we found a match, record it and remove from remaining lists
        if best_match:
            gt_word, ocr_word, distance = best_match
            matches.append((gt_word, ocr_word, distance, 'fuzzy'))
            gt_remaining.pop(best_gt_idx)
            ocr_remaining.pop(best_ocr_idx)
        else:
            # No more matches possible
            break

    # PHASE 3: Mark unmatched words
    # GT words with no match (false negatives)
    for gt_word in gt_remaining:
        matches.append((gt_word, None, None, 'gt_only'))

    # OCR words with no match (false positives)
    for ocr_word in ocr_remaining:
        matches.append((None, ocr_word, None, 'ocr_only'))

    return matches


def create_annotations(word_data, matches, is_ground_truth=True):
    """
    Create annotations for text visualization, preserving original word order.

    Args:
        word_data (list): List of word data dicts with keys: normalized, original, position
        matches (list): Match results from match_words()
        is_ground_truth (bool): True if annotating GT text, False if OCR text

    Returns:
        list: List of annotation dicts with keys: word, match_type, matched_with, edit_distance
    """
    # Build a mapping from normalized words to their match information
    # Handle duplicates by tracking counts
    match_map = {}  # normalized_word -> list of (matched_with, edit_dist, match_type)

    if is_ground_truth:
        for match in matches:
            gt_word, ocr_word, edit_dist, match_type = match
            if match_type in ['exact', 'fuzzy', 'gt_only']:
                if gt_word not in match_map:
                    match_map[gt_word] = []
                match_map[gt_word].append({
                    'matched_with': ocr_word,
                    'edit_distance': edit_dist,
                    'match_type': match_type
                })
    else:
        for match in matches:
            gt_word, ocr_word, edit_dist, match_type = match
            if match_type in ['exact', 'fuzzy', 'ocr_only']:
                if ocr_word not in match_map:
                    match_map[ocr_word] = []
                match_map[ocr_word].append({
                    'matched_with': gt_word,
                    'edit_distance': edit_dist,
                    'match_type': match_type
                })

    # Create annotations in original word order
    annotations = []
    used_counts = {}  # Track which instance of each word we've used

    for word_info in word_data:
        normalized = word_info['normalized']
        original = word_info['original']

        if normalized in match_map:
            # Get the index for this instance of the word
            idx = used_counts.get(normalized, 0)
            used_counts[normalized] = idx + 1

            # Get match info for this instance (if available)
            if idx < len(match_map[normalized]):
                match_info = match_map[normalized][idx]
            else:
                # Fallback if we have more instances than matches
                match_info = match_map[normalized][-1]

            annotations.append({
                'word': original,  # Use original word, not normalized
                'match_type': match_info['match_type'],
                'matched_with': match_info['matched_with'],
                'edit_distance': match_info['edit_distance']
            })
        else:
            # No match found - shouldn't happen but handle gracefully
            annotations.append({
                'word': original,
                'match_type': 'ocr_only' if not is_ground_truth else 'gt_only',
                'matched_with': None,
                'edit_distance': None
            })

    return annotations
